File: parser.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/wait.h>

char *str_replace(const char *orig, const char *rep, const char *with){
	char *result; // the return string
    const char *ins;    // the next insert point
    char *tmp;    // varies
    int len_rep;  // length of rep (the string to remove)
    int len_with; // length of with (the string to replace rep with)
    int len_front; // distance between rep and end of last rep
    int count;    // number of replacements

    // sanity checks and initialization
    if (!orig && !rep)
        return NULL;
    len_rep = strlen(rep);
    if (len_rep == 0)
        return NULL; // empty rep causes infinite loop during count
    if (!with)
        with = "";
    len_with = strlen(with);

    // count the number of replacements needed
    ins = orig;
    for (count = 0; (tmp = strstr(ins, rep)) ; ++count) {
        ins = tmp + len_rep;
    }

    tmp = result = malloc(strlen(orig) + (len_with - len_rep) * count + 1);

    if (!result)
        return NULL;

    // first time through the loop, all the variable are set correctly
    // from here on,
    //    tmp points to the end of the result string
    //    ins points to the next occurrence of rep in orig
    //    orig points to the remainder of orig after "end of rep"
    while (count--) {
        ins = strstr(orig, rep);
        len_front = ins - orig;
        tmp = strncpy(tmp, orig, len_front) + len_front;
        tmp = strcpy(tmp, with) + len_with;
        orig += len_front + len_rep; // move to next "end of rep"
    }
    strcpy(tmp, orig);
    return result;
}

int main(int argc, char *argv[]){
	//Fazer uns error checkings.
	if(argc < 2){
		fprintf(stderr, "missing file.\n");
		return 1;
	}

	//Entrada.
	char *executavel = argv[1];

	//Possível saída.
	FILE *fout = stdout;
	if(argc > 2){
		fout = fopen(argv[2], "w");
	}


	char buff[2048] = {0};
	int pipe_pai[2];
	int pipe_filho[2];
	pid_t pid;

	if(pipe(pipe_pai) || pipe(pipe_filho)){
		perror("pipe");
		return 1;
	}
	
	pid = fork();
	if(pid == -1){
		perror("fork");
		return 1;
	}

	if(pid == 0){
		//Processo filho. Vai ser o gdb.
		//Escreve em pipe_pai[1] e lê em pipe_filho[0].

		//stdin -> pipe_filho[0]
		close(pipe_filho[1]);
		dup2(pipe_filho[0], 0);

		//stdout -> pipe_pai[1]
		close(pipe_pai[0]);
		dup2(pipe_pai[1], 1);

		FILE *err = fopen("/dev/null", "w");
		dup2(fileno(err), 2);

		if(execl("/usr/bin/gdb", "/usr/bin/gdb", "-silent", executavel, (char*) NULL) == -1){
			perror("exec gdb");
			return 1;
		}

	}else{
		//Processo pai. Vai enviar comandos pro gdb.
		//Escreve em pipe_filho[1] e lê em pipe_pai[0].

		close(pipe_filho[0]);
		close(pipe_pai[1]);

        int in = pipe_pai[0];
        int out = pipe_filho[1];

		char comando[128] = "b main\nr\n";

		write(out, comando, strlen(comando) + 1);
		/*sleep(1);
		read(in, buff, sizeof buff);
		printf("saida RUN: %s\n", buff);*/

		for(int i = 0; ; i++){
			strcpy(comando, "\ninfo args\ninfo locals\nn");
			write(out, comando, strlen(comando) + 1);
			//usleep(20000);
			usleep(500000);
			//sleep(1);
			memset(buff, 0, sizeof buff);
			read(in, buff, sizeof buff);
			//fprintf(fout, "NEXT %d:\ncomando: [%s]\n<%s>\n", i, comando, str_replace(buff, "(gdb) ", ""));
			if(strstr(buff, "terminated") != NULL || strstr(buff, "__libc_start_main") != NULL){
				break;
			}
			fprintf(fout, "%s\n============\n", str_replace(buff, "(gdb) ", ""));
			//getchar();
		}

		strcpy(comando, "\nq\ny\nq\n");
		write(out, comando, strlen(comando) + 1);
		/*sleep(1);
		read(in, buff, sizeof buff);
		printf("saida QUIT: %s\n", buff);*/

		//Espera o filho terminar.
		wait(NULL);
		exit(0);
	}
	return 0;
}

File: parser.py
# -*- coding: utf-8 -*-

import os
import sys
import subprocess
from time import sleep
import re
import json

re_identificador = r"[_a-zA-Z][_a-zA-Z0-9]*"
re_funcao = re_identificador + r"\s(?P<funcao>" + re_identificador + r")\(.*\);"
re_instrucao = r"\d+\t.*"
re_frame = r"(?P<var>^" + re_identificador + r" = .+?\n)+"

frames_pattern = re.compile(re_frame, re.DOTALL | re.MULTILINE)
instrucao_pattern = re.compile(re_instrucao)

i = 0
def pbuff(buff, sep = '=', l = 20):
	global i
	print l/2 * sep + " " + str(i) + " " + l/2 * sep + "\n" + buff + "\n" + sep * l
	i += 1

def nmain():
	#Entrada
	executavel = sys.argv[1]

	pipe_pai = os.pipe()
	pipe_filho = os.pipe()

	pid = os.fork()
	if(pid == -1):
		print "Erro no fork"
		exit(1)
	
	if(pid == 0):
		#Processo filho. Vai ser o gdb.
		#Escreve em pipe_pai[1] e lê em pipe_filho[0].

		#stdin -> pipe_filho[0]
		os.close(pipe_filho[1])
		os.dup2(pipe_filho[0], 0)

		#stdout -> pipe_pai[1]
		os.close(pipe_pai[0])
		os.dup2(pipe_pai[1], 1)

		err = open("/dev/null", "w")
		os.dup2(err.fileno(), 2)

		os.execl("/usr/bin/gdb", "/usr/bin/gdb", "-silent", executavel)
	else:
		#Processo pai. Vai enviar comandos pro gdb.
		#Escreve em pipe_filho[1] e lê em pipe_pai[0].

		os.close(pipe_filho[0])
		os.close(pipe_pai[1])

		pin = pipe_pai[0]
		pout = pipe_filho[1]

		def interage(comando, tempo_espera = 0.06, buffer_size = 2048):
			os.write(pout, comando + "\n")
			sleep(tempo_espera)
			if(buffer_size > 0):
				return os.read(pin, buffer_size)

		#Ignorando o lixo inicial.
		sleep(0.1)
		buff = os.read(pin, 2048)

		#Dados necessários
		funcoes = []
		frames = []
		instrucoes = ["main"]

		#Pegando os nomes das funções.
		buff = interage("info fun")

		results = re.finditer(re_funcao, buff)
		for match in results:
			funcao = match.group('funcao')
			funcoes.append(funcao)
			interage("b " + funcao, 0, 0)
		
		buff = interage("r", 0.5)
		results = instrucao_pattern.finditer(buff)
		for match in results:
			instrucoes.append(match.group(0))
		
		pbuff(buff)
		buff = interage("info args\ninfo locals").replace("(gdb) ", "").replace("No arguments.\n", "")
		
		while(not ("__libc_start_main" in buff or "SIGSEGV" in buff)):
			buff = interage("info args\ninfo locals").replace("(gdb) ", "").replace("No arguments.\n", "")
			pbuff(buff)
			buff = interage("n").replace("(gdb) ", "")
			pbuff(buff)

		buff = interage("q\n\y")

		print "=" * 20 + "\n" + buff + "\n" + "=" * 20

		os.wait()
		print funcoes
		exit(0)


def main():
	executavel = sys.argv[1]

	requisitos = []
	frames = []
	instrucoes = ['main']

	saida = subprocess.check_output(["./parser", executavel])

	re_frame = r"(?P<var>^[a-zA-z_]+ = .+?\n)+"
	results = re.finditer(re_frame, saida, re.DOTALL | re.MULTILINE)
	for match in results:
		frames.append(match.group(0).split('\n')[:-1])

	re_frame = r"\d+\t.*"
	results = re.finditer(re_frame, saida)
	for match in results:
		#instrucoes.append(match.group(0).split('\t')[0])
		instrucoes.append(match.group(0))
	
	for i in range(1, len(frames)):
		for j in range(len(frames[i])):
			if(frames[i][j] != frames[i - 1][j]):
				#requisitos.append({""})
				var, valor = frames[i][j].split(' = ')
				if 'char' in instrucoes[i]:
					valor = valor.split(' ')[1]
				requisitos.append({"Nome da Variavel": var, "Valor": valor, "linha": instrucoes[i].split('\t')[0]})
	
	sys.stdout.write(json.dumps(requisitos))
	sys.stdout.flush()

if __name__ == "__main__":
	nmain()
